Added max and min circuit

Shdorsh opened the pull request:
This now has a max and min circuit which checks all values and returns the highest/lowest of them. Tested and working.

[Changelogs]: # This now has a max and min circuit which checks all values and returns the highest/lowest of them. Tested and working.

:cl: Arithmetics plus
add: Min circuit
add: Max circuit
/:cl:

[why]: # Because there seemed to be no way to do a simple min or max function on multiple variables without making huge and messy wiring.

Shdorsh committed:
Added max and min circuit

This now has a max and min circuit which checks all values and returns the highest/lowest of them. Tested and working.

Hate9 commented:
This seems very useful, but as these accept multiple inputs, wouldn't they fit in better with the other parts if they had small and large variants with more and fewer respective inputs?

Shdorsh commented:
It is based off of the typical arithmetics inputs, so it has the same amount of slots as addition and substraction etc. I thought for consistency's sake, I should go with it. Also helps to keep the code lean.

Hate9 commented:
Fair enough, I haven't used the arithmetic parts, so I didn't notice.

Cyberboss reviewed (changes requested):
If you made these two share a base type you could remove pretty much all the code duplication in do_work

Shdorsh committed:
update: saving lines of code

The more compact min max. you may as well want to try the same with addition and substraction.

PKPenguin321 commented:
Would be nice if this accepted lists

Shdorsh commented:
Did some changes to be more compact.

Cyberboss commented on code/modules/integrated_electronics/subtypes/arithmetic.dm:
```diff
@@ -311,3 +311,34 @@
+	extended_desc = "The highest number is put out. Null is ignored."
+	icon_state = "addition"
+	var/state=0
+	spawn_flags = IC_SPAWN_DEFAULT|IC_SPAWN_RESEARCH
```
New vars should be at the bottom of datum definitions

Also `state` doesn't accurately describe what the var does. Code is free, name it something more verbose like `min_comparison`.

Use `FALSE` instead of `0` as well

Cyberboss commented on code/modules/integrated_electronics/subtypes/arithmetic.dm:
```diff
@@ -311,3 +311,34 @@
+	name = "min circuit"
+	desc = "This circuit sends out the smallest number."
+	extended_desc = "The smallest number is put out. Null is ignored. In case no number is found, 0 is given out."
+	state=1
```
`TRUE` instead of `1`

Cyberboss commented on code/modules/integrated_electronics/subtypes/arithmetic.dm:
```diff
@@ -311,3 +311,34 @@
+				result = I
+			else
+				if(I > result && state == 0) result = I
+				else if(I < result && state == 1) result = I
```
The above 5 lines can be simplified to

```dm
if(result == null || (state && I < result) || (!state && I > result))
    result = I
```
Never put control flow body on the same line outside of macros please

Cyberboss commented on code/modules/integrated_electronics/subtypes/arithmetic.dm:
```diff
@@ -311,3 +311,34 @@
+	for(var/k in 1 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(isnum(I))
+			if(!(isnum(result)))
```
Use early return
```dm
if(!isnum(I))
    continue
...
```

Shdorsh replied:
Seems good now

Shdorsh committed:
Reformated code and improved proc

As you said, added early continue, put the vars low and changed the logic on the result.

Cyberboss commented on code/modules/integrated_electronics/subtypes/arithmetic.dm:
```diff
@@ -311,3 +311,32 @@
+	var/result
+	for(var/k in 1 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(!isnum(I))	continue
```
Split this into two lines please

Cyberboss commented on code/modules/integrated_electronics/subtypes/arithmetic.dm:
```diff
@@ -311,3 +311,32 @@
+	for(var/k in 1 to inputs.len)
+		var/I = get_pin_data(IC_INPUT, k)
+		if(!isnum(I))	continue
+		if(!isnum(result) || (!min_comparision && I > result) || (min_comparision && I < result))
```
Change the `!isnum(result)` to `result == null`

Cyberboss commented on code/modules/integrated_electronics/subtypes/arithmetic.dm:
```diff
@@ -311,3 +311,32 @@
+		if(!isnum(I))	continue
+		if(!isnum(result) || (!min_comparision && I > result) || (min_comparision && I < result))
+			result = I
+	if(!isnum(result))
```
Change the `!isnum(result)` to `result == null`

Shdorsh committed:
Update arithmetic.dm
